### PR TITLE
Cassandra 17169 4.0

### PR DESCRIPTION
**Pull request by bereng**

### CODE REVIEW

**Comment by bereng** [comment]
CI 1k repetitions [here](https://app.circleci.com/pipelines/github/bereng/cassandra/511/workflows/833fc9c6-3c0c-4489-b917-573f6ebe2c27)

File: test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java [hunk 1]
```diff
@@ -93,31 +89,40 @@ public void recomputingSupplierTest() throws Throwable
...
                     catch (Throwable e)
                     {
                         thrown.set(e);
-                        latch.countDown();
                     }
                 }
             });
         }
 
-        latch.await(10, TimeUnit.SECONDS);
```
Comment (bereng) [inline test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java:109]: This latch was doing nothing and always timing out

File: test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java [hunk 2]
```diff
@@ -93,31 +89,40 @@ public void recomputingSupplierTest() throws Throwable
...
-
+        Util.spinAssertEquals(true, () -> {return counter.get() > 1;}, 11);
```
Comment (Gerrrr) [inline test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java]: nit: I'd probably take intelij's suggestion and replace the second argument with `() -> counter.get() > 1` which is a little less verbose.

**Review by Gerrrr — APPROVED** [review]
The patch looks good to me. I was able to reproduce the bug with the instructions you shared in `trunk` and couldn't do it again on this branch. Besides fixing flakiness, it also dramatically speeds up test execution - on my local machine test duration reduced from ~10s to ~10ms.

As a suggestion, in my opinion this test case might benefit from a couple of comments explaining the intention. For example,

```
This test case verifies that recomputing supplier never returns out of order values during concurrent updates and eventually returns the most recent value.
```

File: test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java (hunk 2 above)
Comment (bereng) [reply test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java]: Eclipse complains if I don't use the `return` :shrug:

Comment (Gerrrr) [reply test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java]: Sorry for nitpicking then!

Comment (adelapena) [reply test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java]: I understand that the purpose of this line is ensuring that the tasks above have had time to update `counter`. The previous incarnation of this line awaited for 10s which is probably too much, but I'm wondering if waiting for a single increase of counter is too little. Maybe we can wait until it has been increased more times, for example:
```suggestion
        Util.spinAssertEquals(true, () -> {return counter.get() > 1000;}, 11);
```

File: test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java [hunk 3]
```diff
@@ -93,31 +89,40 @@ public void recomputingSupplierTest() throws Throwable
...
         testExecutor.shutdown();
-        testExecutor.awaitTermination(10, TimeUnit.SECONDS);
-        executor.shutdown();
-        executor.awaitTermination(10, TimeUnit.SECONDS);
+        Assert.assertTrue(testExecutor.awaitTermination(11, TimeUnit.SECONDS));
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java]: Why are we waiting 11 seconds in the calls to `spinAssertEquals` and `awaitTermination`? If it's an arbitrary amount of time we could use a round number like for example 10s.

File: test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java [hunk 4]
```diff
@@ -93,31 +89,40 @@ public void recomputingSupplierTest() throws Throwable
...
             throw new AssertionError(supplier.toString(), thrown.get());
 
-        Assert.assertTrue(counter.get() > 1); // We actually did some work
-        Assert.assertEquals(supplier.get(1, TimeUnit.SECONDS).longValue(), counter.get());
+        Util.spinAssertEquals(true, () -> {
+            try
+            {
+                return supplier.get(1, TimeUnit.SECONDS).longValue() == counter.get();
+            }
+            catch(InterruptedException | ExecutionException | TimeoutException e)
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java]: Super nit: missed space
```suggestion
            catch (InterruptedException | ExecutionException | TimeoutException e)
```

File: test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java [hunk 5]
```diff
@@ -93,31 +89,40 @@ public void recomputingSupplierTest() throws Throwable
...
 
         if (thrown.get() != null)
             throw new AssertionError(supplier.toString(), thrown.get());
 
-        Assert.assertTrue(counter.get() > 1); // We actually did some work
-        Assert.assertEquals(supplier.get(1, TimeUnit.SECONDS).longValue(), counter.get());
+        Util.spinAssertEquals(true, () -> {
+            try
+            {
+                return supplier.get(1, TimeUnit.SECONDS).longValue() == counter.get();
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java]: Nit: we don't need the call to `longValue`:
```suggestion
                return supplier.get(1, TimeUnit.SECONDS) == counter.get();
```

File: test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java [hunk 6]
```diff
@@ -93,31 +89,40 @@ public void recomputingSupplierTest() throws Throwable
...
+            }
+            catch(InterruptedException | ExecutionException | TimeoutException e)
+            {
+                e.printStackTrace();
+                return false;
+            }
+        }, 11);
+
+        executor.shutdown();
+        Assert.assertTrue(executor.awaitTermination(11, TimeUnit.SECONDS));
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java]: Maybe we could check that once the executor is stopped the supplier and the counter still have the same value?
```
Assert.assertEquals((long) supplier.get(0, TimeUnit.MILLISECONDS), counter.get());
```

File: test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java (hunk 2 above)
Comment (adelapena) [reply test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java]: > nit: I'd probably take intelij's suggestion and replace the second argument with () -> counter.get() > 1 which is a little less verbose.

I'd also go with the concise format, `() -> counter.get() > 1`. I haven't seen any rule about this in our code style, so I guess that the IDE warning depends on random defaults not covered by the project's code style. The `eclipse-warnings` Ant target seems to accept both formats. Anyway, I haven't found anything on the codebase containing the pattern `) -> {return`, while there are numerous usages of the compact format. So, we can probably assume that that's the untold convention.

Comment (bereng) [reply test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java]: Mmmm I must have made a typo before bc I tested this now again and it doesn't complain :shrug: Changed.

Comment (bereng) [reply test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java]: On the waiting of 11s it is a left over of a previous approach of 10 threads with a 1s sleep bla bla tes I was doing. So I was waiting 11s to give time to all 10 threads. Makes no sense now and 10 reads better, right.

On the purpose of that line giving 10s of time to recompute I am not sure. Why not just sleep? Why wait for an exception that never happens? I never grasped fully that approach.

File: test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java (hunk 3 above)
Comment (bereng) [reply test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java]: Part of a left over from a previous approach where I had 10 threads sleeping 1s, hence I was waiting 11s. We can change it to 10 which reads better.

**Comment by bereng** [comment]
Review comments applied. Survives 1,5K local runs. If we +1 I'll run 1K on circle.

File: test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java [hunk 7]
```diff
@@ -93,31 +91,41 @@ public void recomputingSupplierTest() throws Throwable
...
                         thrown.set(e);
-                        latch.countDown();
                     }
                 }
             });
         }
 
-        latch.await(10, TimeUnit.SECONDS);
-
+        Util.spinAssertEquals(true, () -> counter.get() > 1000, 10);
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java]: I think that the purpose of the latch was waiting for ten seconds to give time to the tasks to interact with the tested supplier, but abort immediately if an exception was thrown. Waiting for a fixed counter value of 1000 means a shorter wait in most of the cases, and less chances for races in both the test and the tested class.

File: test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java [hunk 8]
```diff
@@ -93,31 +91,41 @@ public void recomputingSupplierTest() throws Throwable
...
+            catch (InterruptedException | ExecutionException | TimeoutException e)
+            {
+                e.printStackTrace();
+                return false;
+            }
+        }, 10);
+
+        executor.shutdown();
+        Assert.assertTrue(executor.awaitTermination(10, TimeUnit.SECONDS));
+        Assert.assertEquals((long) supplier.get(0, TimeUnit.MILLISECONDS), counter.get());
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java]: Unfortunately this hits three timeouts in a CI round with 2000 iterations, [here](https://app.circleci.com/pipelines/github/adelapena/cassandra/1180/workflows/1a64ccfc-1cb7-4b36-8e42-6489f1442af2/jobs/10603/tests). I guess the reason is that `executor` hasn't finished yet. Increasing that timeout from 0 to 10 sec doesn't fix the problem, as it's shown by [this other run](https://app.circleci.com/pipelines/github/adelapena/cassandra/1181/workflows/86e29e09-b523-40c8-b2ea-8df81be5910c/jobs/10606). Also, the latter run hits a different, earlier failure due to a timeout in the supplier reader task.

File: test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java (hunk 7 above)
Comment (bereng) [reply test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java]: Would we prefer to give it 10s again then or is 1K iterations enough? :thinking:

Comment (adelapena) [reply test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java]: I don't really have a preference, I was just commenting about why the latch was there. The `spinAssert` is briefer  and probably easier to understand so I guess I'd lean towards it. As for the 1k iterations, that's probably as arbitrary as the 10 seconds, and still it doesn't prevent us from hitting the failures mentioned below, so I guess we can keep it.

**Comment by bereng** [comment]
Moving to 4K repetitions bc I managed a successful 2K one

File: test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java (hunk 8 above)
Comment (bereng) [reply test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java]: 4K green pass [here](https://app.circleci.com/pipelines/github/bereng/cassandra/521/workflows/ea388d7d-4b87-486c-ae1c-cb9c286d8dec/jobs/4639). The 'earlier' error you hit was a 1s future `get()` I moved to 5s and everything else to 30s to accommodate for 25s (complete finger in the air) in the latest commit.

Comment (adelapena) [reply test/unit/org/apache/cassandra/utils/RecomputingSupplierTest.java]: Looks good to me, +1.

**Review by adelapena — APPROVED** [review]

**Comment by bereng** [comment]
Squashed before merge

**Pull request closed without merging**